_posts/2015-05-22-Notifying-you-of-changes-to-notifications.md

```diff
@@ -0,0 +1,309 @@
+  var icon = '/images/icon-192x192.png';    
+  var tag = 'simple-push-demo-notification-tag';  
+  var data = {  
+    doge: {  
+        wow: 'such amaze notification data'  
+    }  
+  };
+
+  event.waitUntil(    
+    self.registration.showNotification(title, {    
+      body: body,    
+      icon: icon,    
+      tag: tag,  
+      data: data    
+    })    
+  );    
+});
+{% endhighlight %}
+
+Means we can get the information in the notificationclick event:
+
+{% highlight javascript %}
+self.addEventListener('notificationclick', function(event) {  
+  var doge = event.notification.data.doge;  
+  console.log(doge.wow);  
+});
+{% endhighlight %}
+
+Before this, you had to stash data in IndexDB or put something on the end of the 
+icon URL - eek.
+
+# ServiceWorkerRegistration.getNotifications()
+
+One common request from developers working on push notifications is to have 
+better control over the notifications that they display.
+
+An example use case would be a chat application where a user sends multiple 
+messages and the recipient displays multiple notifications. Ideally the web app 
+would be able to notice you have several notifications which haven't been viewed 
+and collapse them down into a single notification.
+
+Without getNotifications() the best you can do is replace the previous 
+notification with the latest message. With getNotifications(), you can 
+"collapse" the notifications if a notification is already displayed - leading to 
+a much better user experience.
+
+<p style="text-align: center; text-align: center; margin: 2em 0 2em 0;">
+  <img style="max-width: 100%; height: auto;" src="{% asset_path 2015-05-24-notifying-you-of-notification-changes/notifications-all.png %}" alt="Example of grouping notifications together" width="601" height="262" />
+</p>
+
+The code to do this is relatively simple. Inside your push event, call 
+ServiceWorkerRegistration.getNotifications() to get an array of current 
+Notifications and from there decide the right behaviour, whether that's 
+collapsing all notifications or by using the Notification.tag.
+
+{% highlight javascript %}
+function showNotification(title, body, icon, data) {  
+  var notificationOptions = {  
+    body: body,  
+    icon: icon ? icon : 'images/touch/chrome-touch-icon-192x192.png',  
+    tag: 'simple-push-demo-notification',  
+    data: data  
+  };  
+  if (self.registration.showNotification) {  
+    self.registration.showNotification(title, notificationOptions);  
+    return;  
+  } else {  
+    new Notification(title, notificationOptions);  
+  }  
+}
+
+self.addEventListener('push', function(event) {  
+  console.log('Received a push message', event);
+
+  // Since this is no payload data with the first version  
+  // of Push notifications, here we'll grab some data from  
+  // an API and use it to populate a notification  
+  event.waitUntil(  
+    fetch(API\_ENDPOINT).then(function(response) {  
```

Comment:
Pretty much every use of push today will perform a fetch, but what it's fetching isn't relevant.